### PR TITLE
planner/core: change agg cost factor

### PR DESCRIPTION
**Pull request by hanfei1991**
### What problem does this PR solve?

Problem Summary:

Right now the cost factor of only-group-by (always transformed from `select distinct ....`) statements is 1.0, leading to not pushing down agg in pseodu stats senerios. However, the cost factor is not reasonable to set 1.0 while the first-row's cost factor is only 0.1. This results in some unexpected behaviours, such as:

```
mysql> explain select a, b from t1 group by a;
+---------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------+
| id                        | estRows  | task      | access object | operator info                                                                                 |
+---------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------+
| HashAgg_9                 | 8000.00  | root      |               | group by:test.t1.a, funcs:firstrow(test.t1.a)->test.t1.a, funcs:firstrow(Column#5)->test.t1.b |
| └─TableReader_10          | 8000.00  | root      |               | data:HashAgg_5                                                                                |
|   └─HashAgg_5             | 8000.00  | cop[tikv] |               | group by:test.t1.a, funcs:firstrow(test.t1.b)->Column#5                                       |
|     └─TableFullScan_8     | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                |
+---------------------------+----------+-----------+---------------+-----------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> explain select a from t1 group by a;
+--------------------------+----------+-----------+---------------+----------------------------------------------------------+
| id                       | estRows  | task      | access object | operator info                                            |
+--------------------------+----------+-----------+---------------+----------------------------------------------------------+
| HashAgg_7                | 8000.00  | root      |               | group by:test.t1.a, funcs:firstrow(test.t1.a)->test.t1.a |
| └─TableReader_12         | 10000.00 | root      |               | data:TableFullScan_11                                    |
|   └─TableFullScan_11     | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                           |
+--------------------------+----------+-----------+---------------+----------------------------------------------------------+
3 rows in set (0.00 sec)
```

It's hard to say which one is better because of the lack of stats info. But at least keeping same behaviours will not cause perils. In some MPP senarios, such as every complicated sql, the plans closer to the root (TiDB) might miss the stats info and choose 1-phase aggregation. Considering 2-phase agg is better than 1-phase in most cases, I think this change is beneficial.

However, we still need some experiments and reliable tests to calibrate the cost factors. **This work should be done in the near future.** 

### What is changed and how it works?

What's Changed:
change the factor of only-group-by agg from 1.0 to 0.1, which is same as first-row.

Tests 

- Unit test. Over 100 lines of relavent tests are changed. Some tests involed with in-subquery will generate a "distinct" group by. I have checked the test changes and made sure they are safe.

### Release note 

- planner/core: change agg cost factor

### CODE REVIEW

File: planner/core/physical_plans.go [hunk 1]
```diff
@@ -1058,7 +1058,7 @@ func (p *basePhysicalAgg) getAggFuncCostFactor() (factor float64) {
 		}
 	}
 	if factor == 0 {
-		factor = 1.0
+		factor = 0.1
```
Comment (winoros) [inline planner/core/physical_plans.go:1061]: Maybe you could add some comments and directly use `aggFuncFactor[firstrow]`.

Comment (hanfei1991) [reply planner/core/physical_plans.go:1061]: ok, using `aggFuncFactor[firstrow]` is a good idea!

File: planner/core/physical_plans.go [hunk 2]
```diff
@@ -1058,7 +1058,11 @@ func (p *basePhysicalAgg) getAggFuncCostFactor() (factor float64) {
 		}
 	}
 	if factor == 0 {
-		factor = 1.0
+		if isMPP {
```
Comment (winoros) [inline planner/core/physical_plans.go:1061]: Add comments and TODO for this if branch.

Comment (hanfei1991) [reply planner/core/physical_plans.go:1061]: done

File: planner/core/physical_plans.go [hunk 3]
```diff
@@ -1058,7 +1058,15 @@ func (p *basePhysicalAgg) getAggFuncCostFactor() (factor float64) {
...
+			// The default factor 1.0 will lead to 1-phase agg in pseudo stats settings.
+			// But in mpp cases, 2-phase is more usual. So we change this factor.
+			// TODO: This is still a little tricky and might cause regression. We should
+			// calibrate these factors and polish our cost model in the future.
+			factor = aggFuncFactor[ast.AggFuncFirstRow]
```
Comment (qw4990) [inline planner/core/physical_plans.go:1066]: Please add some test cases.

Comment (qw4990) [reply planner/core/physical_plans.go:1066]: Oh it can be covered by some existing test cases.

**Review by qw4990 — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**